AK/GenericLexer.cpp

```diff
@@ -0,0 +1,335 @@
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume and return characters until the string `stop` is found
+// The `stop` string is ignored, as it is user-defined
+StringView GenericLexer::consume_until(const char* stop)
+{
+    size_t start = m_index;
+    while (!is_eof() && !next_is(stop))
+        m_index++;
+    size_t length = m_index - start;
+
+    ignore(__builtin_strlen(stop));
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume and return characters until `condition` return true
+StringView GenericLexer::consume_until(Condition condition)
+{
+    size_t start = m_index;
+    while (!is_eof() && !condition(peek()))
+        m_index++;
+    size_t length = m_index - start;
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume a string surrounded by single or double quotes
+// The returned StringView does not include the quotes
+StringView GenericLexer::consume_quoted_string()
+{
+    if (!is_quote(peek()))
+        return {};
+
+    char quote_char = consume();
+    size_t start = m_index;
+    while (!is_eof() && peek() != quote_char)
+        m_index++;
+    size_t length = m_index - start;
+
+    if (peek() != quote_char) {
+        m_index = start - 1; // Restore the index in case the string is unterminated
+        return {};
+    }
+
+    ignore();
+
+    return m_input.substring_view(start, length);
+}
+
+// Ignore a number of characters (1 by default)
+void GenericLexer::ignore(size_t count)
+{
+    count = min(count, m_input.length() - m_index);
+    m_index += count;
+}
+
+// Ignore characters while `condition` returns true
+void GenericLexer::ignore_while(Condition condition)
+{
+    while (!is_eof() && condition(peek()))
+        m_index++;
+}
+
+// Ignore characters until `stop` is peek'd
+// The `stop` character is ignored as it is user-defined
+void GenericLexer::ignore_until(char stop)
+{
+    while (!is_eof() && peek() != stop)
+        m_index++;
+
+    ignore();
+}
+
+// Ignore characters until the string `stop` is found
+// The `stop` string is ignored, as it is user-defined
+void GenericLexer::ignore_until(const char* stop)
+{
+    while (!is_eof() && !next_is(stop))
+        m_index++;
+
+    ignore(__builtin_strlen(stop));
+}
+
+// Ignore characters until `condition` return true
+// We don't skip the stop character as it may not be a single value
+void GenericLexer::ignore_until(Condition condition)
+{
+    while (!is_eof() && !condition(peek()))
+        m_index++;
+}
+
+
+bool is_control(char c)
+{
+    return (c >= 0 && c <= 31) || c == 127;
+}
+
+bool is_whitespace(char c)
+{
+    return (c >= '\t' && c <= '\r') || c == ' ';
+}
+
+bool is_lowercase(char c)
+{
+    return c >= 'a' && c <= 'z';
+}
+
+bool is_uppercase(char c)
+{
+    return c >= 'A' && c <= 'Z';
+}
+
+bool is_digit(char c)
+{
+    return c >= '0' && c <= '9';
+}
+
+bool is_punctuation(char c)
```

Comment:
Do we have the [c functions](http://www.cplusplus.com/reference/cctype/) avaliable for this?
ispunct, islower, isupper, isdigit, iscntrl, etc

Reply:
Yep, I know, however when I tried to include the ctype.h I ended with C/C++ linkage errors. I think AK is used both inside the Kernel and Serenity, the OS (there are some `#if defined(KERNEL)` hanging around so I guess there are linkage differences). I wasn't really sure how to correct this

Reply:
Sorry for the late reply. Here is what I ended up with when trying to include the ctype.h:
![image](https://user-images.githubusercontent.com/10747470/86614708-b4b47d00-bfb3-11ea-8376-23349bf85bbc.png)
Trying to fix this (if it should) would have been out of scope from my PR so I went with what you pointed out instead.

Reply:
Woah, that seems very wrong. LibC/string.h should be entirely `extern "C"` because of the __BEGIN_DECLS wrapping all the methods.